crowbar_framework/lib/crowbar/backup/restore.rb

```diff
@@ -23,26 +23,118 @@ def initialize(backup)
         @backup = backup
         @data = @backup.data
         @version = @backup.version
+        @status = {}
       end
 
       def restore
-        actions = [
-          :restore_crowbar,
-          :run_installer,
-          :restore_chef_keys,
-          :restore_chef,
-          :restore_database
-        ]
-        actions.each do |component|
-          ret = send(component)
-          return ret unless ret == true
-        end
-
-        { status: :ok, msg: "" }
+        self.class.restore_steps_path.delete if self.class.restore_steps_path.exist?
+
+        Thread.new do
+          self.class.steps.each do |component|
+            set_step(component)
+            send(component)
+            set_success && self.class.restore_steps_path.delete if component == :restore_database
+            return @status && set_failed && Thread.exit if any_errors?
+          end
+        end
+      end
+
+      class << self
+        def status
+          {
+            steps: steps_done,
+            success: success?,
+            failed: failed?,
+            restoring: restoring?
+          }
+        end
+
+        def steps
+          [
+            :restore_crowbar,
+            :run_installer,
+            :restore_chef_keys,
+            :restore_chef,
+            :restore_database
+          ]
+        end
+
+        def restore_steps_path
+          install_dir_path.join("restore_steps")
+        end
+
+        def install_dir_path
+          Pathname.new("/var/lib/crowbar/install")
+        end
+
+        def failed_path
+          install_dir_path.join("crowbar-restore-failed")
+        end
+
+        def success_path
+          install_dir_path.join("crowbar-restore-ok")
+        end
+
+        def install_dir_path
```

Comment:
Lint/DuplicateMethods: Method `Crowbar::Backup::Restore.install_dir_path` is defined at both /app/crowbar_framework/lib/crowbar/backup/restore.rb:66 and /app/crowbar_framework/lib/crowbar/backup/restore.rb:78.